tests/miral/output_updates.cpp

```diff
@@ -0,0 +1,125 @@
+        .WillOnce(Invoke([&](Output const& output){ output_a = output; }));
+
+    set_outputs({display_area_a});
+
+    Mock::VerifyAndClearExpectations(window_manager_policy);
+    EXPECT_THAT(output_a.value().extents(), Eq(display_area_a));
+}
+
+TEST_F(OutputUpdates, policy_notified_of_multiple_outputs)
+{
+    std::experimental::optional<Output> output_a;
+    std::experimental::optional<Output> output_b;
+
+    EXPECT_CALL(*window_manager_policy, advise_output_create(_))
+        .WillOnce(Invoke([&](Output const& output){ output_a = output; }))
+        .WillOnce(Invoke([&](Output const& output){ output_b = output; }));
+
+    set_outputs({display_area_a, display_area_b});
+
+    Mock::VerifyAndClearExpectations(window_manager_policy);
+    EXPECT_THAT(output_a.value().extents(), Eq(display_area_a));
+    EXPECT_THAT(output_b.value().extents(), Eq(display_area_b));
+    EXPECT_FALSE(output_a.value().is_same_output(output_b.value()));
+}
+
+TEST_F(OutputUpdates, policy_notified_of_output_update)
+{
+    std::experimental::optional<Output> output_initial;
+    std::experimental::optional<Output> output_original;
+    std::experimental::optional<Output> output_updated;
+
+    EXPECT_CALL(*window_manager_policy, advise_output_create(_))
+        .WillOnce(Invoke([&](Output const& output){ output_initial = output; }));
+
+    set_outputs({display_area_a});
+
+    Mock::VerifyAndClearExpectations(window_manager_policy);
+
+    EXPECT_CALL(*window_manager_policy, advise_output_update(_, _))
+        .WillOnce(Invoke([&](Output const& updated, Output const& original){
+            output_original = original;
+            output_updated = updated;
+        }));
+
+    set_outputs({display_area_b});
+
+    Mock::VerifyAndClearExpectations(window_manager_policy);
+    EXPECT_TRUE(output_initial.value().is_same_output(output_original.value()));
+    EXPECT_TRUE(output_original.value().is_same_output(output_updated.value()));
+
+    EXPECT_THAT(output_initial.value().extents(), Eq(display_area_a));
+    EXPECT_THAT(output_original.value().extents(), Eq(display_area_a));
+    EXPECT_THAT(output_updated.value().extents(), Eq(display_area_b));
+}
+
+TEST_F(OutputUpdates, policy_notified_of_output_delete)
+{
+    std::experimental::optional<Output> output_a;
+    std::experimental::optional<Output> output_b;
+    std::experimental::optional<Output> output_b_deleted;
+
+    EXPECT_CALL(*window_manager_policy, advise_output_create(_))
+        .WillOnce(Invoke([&](Output const& output){ output_a = output; }))
+        .WillOnce(Invoke([&](Output const& output){ output_b = output; }));
+
+    set_outputs({display_area_a, display_area_b});
+
+    Mock::VerifyAndClearExpectations(window_manager_policy);
+
+    EXPECT_CALL(*window_manager_policy, advise_output_delete(_))
+        .WillOnce(Invoke([&](Output const& output){ output_b_deleted = output; }));
+
+    set_outputs({display_area_a});
+
+    Mock::VerifyAndClearExpectations(window_manager_policy);
```

Comment:
Calling Mock::VerifyAndClearExpectations() multiple times "smells" of testing multiple things in one test. At the very least there should be a comment on why it is called at all.

Reply:
Before the rest of the test can happen, we need to make sure the expected functions have been called or else variables we expect to hold values will be unset. If Mir was broken in such a way the calls were not made, I think it would be more helpful to see the error that those functions were not called, than see an error that the values they set are not what is expected. I have dropped some comments in.